From 6f26331d24745e9c30c37720fc3061b99657c1e0 Mon Sep 17 00:00:00 2001
From: Venkataramana Kotakonda <venkataramana.kotakonda@intel.com>
Date: Mon, 12 Jan 2015 12:05:22 +0530
Subject: [PATCH 1/1] XPWR: Fixed charger detection error during boot.

If system booted with charger connected, the xpwr detects SDP as DCP
as USB data lines are connected to soc. hence, driver should re-initiate
the charger detection. This patch provide this fix.

Change-Id: I53edce3eaca52717181bfa9eeef9474c0c1f5a77
Tracked-On: https://jira01.devtools.intel.com/browse/AKP-843
Signed-off-by: Venkataramana Kotakonda <venkataramana.kotakonda@intel.com>
---
 drivers/extcon/extcon-dc-pwrsrc.c | 16 +++++++---------
 1 file changed, 7 insertions(+), 9 deletions(-)

diff --git a/drivers/extcon/extcon-dc-pwrsrc.c b/drivers/extcon/extcon-dc-pwrsrc.c
index ffd66fe..eec961f 100644
--- a/drivers/extcon/extcon-dc-pwrsrc.c
+++ b/drivers/extcon/extcon-dc-pwrsrc.c
@@ -82,7 +82,7 @@
 #define PWRSRC_IRQ_CFG_MASK		0x1C
 
 #define DC_BC12_IRQ_CFG_REG		0x45
-#define BC12_IRQ_CFG_MASK		0x3
+#define BC12_IRQ_CFG_MASK		0x2
 
 #define DC_XPWR_CHARGE_CUR_DCP		2000
 #define DC_XPWR_CHARGE_CUR_CDP		1500
@@ -487,20 +487,18 @@ static int dc_xpwr_pwrsrc_probe(struct platform_device *pdev)
 
 	/* Unmask VBUS interrupt */
 	intel_soc_pmic_writeb(DC_PWRSRC_IRQ_CFG_REG, PWRSRC_IRQ_CFG_MASK);
+	intel_soc_pmic_clearb(DC_BC_GLOBAL_REG, BC_GLOBAL_RUN);
 	if (info->pdata->en_chrg_det) {
 		/* unmask the BC1.2 complte interrupts */
 		intel_soc_pmic_writeb(DC_BC12_IRQ_CFG_REG, BC12_IRQ_CFG_MASK);
 		/* enable the charger detection logic */
 		intel_soc_pmic_setb(DC_BC_GLOBAL_REG, BC_GLOBAL_RUN);
-	}
-
-	if (info->pdata->en_chrg_det)
-		ret = handle_chrg_det_event(info);
-	else
+	} else {
 		ret = handle_pwrsrc_event(info);
-	if (ret < 0)
-		dev_warn(&info->pdev->dev, "error in PWRSRC evt handling\n");
-
+		if (ret < 0)
+			dev_warn(&info->pdev->dev,
+				"error in PWRSRC evt handling\n");
+	}
 	return 0;
 
 intr_reg_failed:
-- 
1.9.1

